### DISCUSSION TITLE
Bootstrap diagram

### DISCUSSION THREAD

**Discussion by modship** [massalabs/massa#3609] [General]
Here is the current connection diagram for the bootstrap refactor of #3554  : 

@Ben-PH 

Update with the fix of updater loop (deprecated) : 

![image](https://user-images.githubusercontent.com/9607217/221876600-92a8349c-6bec-4715-92eb-b645553de06c.png)

deprecated :
![Diagramme sans nom (1)](https://user-images.githubusercontent.com/9607217/221516168-2c6b42ae-7057-4e18-bade-9e5cf5bfdac6.jpg)

**Comment by AurelienFT** [comment]
I think for me the main loop/listener should be either regroup (as there is only a listener management in the main loop) or the listener should create the thread himself. I find this message passing a bit too big overhead. The verifications could be done in the thread of the client or just before as it's read only with a `RwLock` it will not prevent to have differents threads checking the same data.

**Reply by Ben-PH** [reply]
I agree. The issue, though, is the async nature of the listener. Not just that it _is_ async, but also the _how_ it's setup.

Once it becomes sync, putting it in the main loop, without the channel, will _Just Work_^tm

**Reply by AurelienFT** [reply]
Ok !

**Reply by Ben-PH** [reply]
So it wasn't quite as trivial as I originally thought: what we _want_ is to be able to do this:
```rust
select! {
    recv(stopper) -> _ => todo!("stop the system"),
    recv_from_os(listener.accept()) -> connection => todo!("handle an acceptance"),
}
```
essentially, to use the crossbeam select system to do the stop/handle branching.

The problem is that crossbeam select only works on channels. We need a way to wake up when _either_ is ready. Because a basic `TcpListener::accept()` causes the OS to block the whole thread, that would normally block receiving on the channel as well.

The solution is to set the listener to non-blocking, and do a polling select:

```rust
listener.set_nonblocking(true);
let non_block_accept = listener.accept();
if non_block_accept.is_err() {
    if non_block_accept.unwrap().kind() == io::Error::ErrorKind::WouldBlock {
        // handle this branch of code the same way that crossbeam handles Receiver::try_receive() returning the equivilent to "would block"
    }
    // ...
}
```

So I've added a helper (first impl. Later should be better) which is basically a re-implementation of crossbeams select internals: https://github.com/massalabs/massa/commit/f561249b33ed7c8e31ba4f5941a1ba4d2df8f85a#diff-2f2a0d279d4c015c632b27bdc7bd2fb62cf753fa8811dead19aadc3133666156R395-R423

https://docs.rs/crossbeam-channel/0.5.6/src/crossbeam_channel/select.rs.html#176 for the original

https://docs.rs/crossbeam/latest/crossbeam/utils/struct.Backoff.html for the backoff example that is much simpler that the linked commit referred to.

This allows the listener thread to be dropped entirely.

here is a MVP:

```rust
use std::{
    net::{SocketAddr, TcpListener, TcpStream},
    time::{Duration, Instant},
};

use crossbeam::{channel::Receiver, utils::Backoff};

fn main() {
    let mut listener = TcpListener::bind("localhost:8080").unwrap();
    listener.set_nonblocking(true);
    let (stop_tx, mut stop_rx) = crossbeam::channel::bounded::<()>(2);
    let sleep_stopper = std::thread::spawn(move || {
        std::thread::sleep(Duration::from_secs(20));
        dbg!("awake");
        stop_tx.send(());
    });
    loop {
        if let Err(_) = dbg!(select(&mut stop_rx, &mut listener)) {
            break;
        }
    }
}
fn select(
    stop_rx: &mut Receiver<()>,
    listener: &mut TcpListener,
) -> Result<(TcpStream, SocketAddr), ()> {
    'select: loop {
        let backoff = Backoff::new();
        'backoff: loop {
            if stop_rx.try_recv().is_ok() {
                dbg!("returning because stopped");
                return Err(());
            }

            let try_accept = listener.accept();
            match try_accept {
                Ok(good_conn) => {
                    return Ok(good_conn);
                }
                Err(e) => {
                    if e.kind() == std::io::ErrorKind::WouldBlock {
                        dbg!("zzz");
                        backoff.snooze();
                    } else {
                        panic!("try_accept got a bad error");
                    }
                }
            }
        }
    }
}
```

run that, and you have 20 seconds to run `telnet 127.0.0.1:8080`

**Reply by Ben-PH** [reply]
Essentially, the idea is we use the same approach crossbeam uses to poll a channel as ready, to poll the OS as ready to provide a connection.

...might be good for an upstream commit: add the ability to add would-block sys-calls to `crossbeam::select!` through a feature flag named "os_recv" or something...

**Reply by Ben-PH** [reply]
This solution is probbly not that great, actually. polling is still a system-call, which invokes a context switch, which is... not ideal under this model: putting it in a back-off spin-lock causes a flood of sys-calls, whereas a thread-dedicated listener with channel doesn't spam expensive syscalls.

**Reply by AurelienFT** [reply]
I had the same problem for network refactoring and I ended up using Mio for this : https://github.com/tokio-rs/mio#usage it works well : here is what i've done (less readable than their readme): https://github.com/massalabs/PeerNet/blob/eed4edcdce6ed3a2f9721fcf76268f6dbc34cfa4/src/transports/tcp.rs

**Reply by Ben-PH** [reply]
I love the idea of that. I'm just curious about how it's done in a way that doesn't spam syscalls, and doesn't block the polling of the stopper.

Assuming that's done, it's pretty ideal. I wouldn't be surprised if that assumption is upheld, but I don't want to fly in blindly.

The other approach I was going to try, is to use `std::sync::Condvar`, and use that to manage thread-progress. Mio will be much nicer though.

**Reply by Ben-PH** [reply]
...I've had a look into it. I took the tcp server examples and added some printlns... no spamming. Will use mio

**Reply by modship** [reply]
> ...I've had a look into it. I took the tcp server examples and added some printlns... no spamming. Will use mio

Nice 👌🏻

**Reply by AurelienFT** [reply]
It's used by tokio so I think it's optimize but didn't go in depth

**Reply by Ben-PH** [reply]
Yeah, I cloned the repo, added some printlns before the `syscall!`s, and saw that we're good.

I'm surprised that crossbeam doesn't bring it in to play, but that's another discussion.
